### ISSUE TITLE
Consider nox instead of tox+make+shell script for automation

### ISSUE THREAD

**Issue by tiran** [instructlab/instructlab#1194]
**Is your feature request related to a problem? Please describe.**
At the moment, instructlab is using a mix of tox, Makefile, shell scripts, and pre-commit for automation.

- [tox](https://tox.wiki) is used create virtual envs and drive testing, linting, and code formatting
- Makefile is used for container images, to run some tox targets, run container for markdown validation, and more
- tox uses `scripts/ruff.sh` to run Ruff's checker and formatter
- pre-commit is used by some to perform the same code and format checkers
- The `README.md` also contains a bunch of rather complex instructions how to create a virtual env and install a PyTorch stack for CPU, CUDA, ROCm, Apple MPS, Apple on Intel, and so on.

**Describe the solution you'd like**
Nox would allow us to consolidate tox, Makefile, and shell script into a single system and single file. From Nox's documentation:

> [nox](https://nox.thea.codes/en/stable/index.html) is a command-line tool that automates testing in multiple Python environments, similar to [tox](https://tox.readthedocs.org/). Unlike tox, Nox uses a standard Python file for configuration.

In other words, Nox is a powerful alternative for Tox. Tox uses a domain specific language (ini files), which limits it and makes it less flexible. Nox uses Python code to define commands. It has the full power of Python at its disposal. This makes Nox more versatile.

We could replace all of tox and most of make with Nox. On top of that, we can introduce a command to create a virtual env for the user with correct build and pip flags for GPUs, MPS, and so on. PyBaMM is a good [example](https://github.com/pybamm-team/PyBaMM/blob/8823b40dfc55a9339f3a32959ac79d4347ad75ca/noxfile.py#L113-L132) for dev env creation.

**Additional context**
* upstream docs: https://nox.thea.codes/en/stable/
* https://sethmlarson.dev/nox-pyenv-all-python-versions
* https://hynek.me/articles/why-i-like-nox/

**Comment by leseb** [comment]
This is an interesting proposal, thanks for the writeup!
I have a few observations:

>We could replace all of tox and most of make with Nox

What part of the Make file would move to Nox?

> On top of that, we can introduce a command to create a virtual env for the user with correct build and pip flags for GPUs, MPS, and so on. PyBaMM is a good [example](https://github.com/pybamm-team/PyBaMM/blob/8823b40dfc55a9339f3a32959ac79d4347ad75ca/noxfile.py#L113-L132) for dev env creation.

So this cannot be done with Tox?

>tox uses scripts/ruff.sh to run Ruff's checker and formatted

So will `ruff.sh` be rewritten in Python and included in Nox?

Even though consolidation sounds good, I don't see a strong argument or urgent need for this switch. It looks like it would take quite a bit of work, and so far, I haven't seen anything we can't do with Tox that Nox would fix.
All that being said I'm opposed to this, especially if you want to tackle it and have the capacity. I just want to make sure we’re not fixing what isn’t broken or always jumping to refactor with the latest cool tool. I know that’s not what you’re suggesting, I’m just being a bit cautious, that’s all! :)

Thanks!

**Comment by alinaryan** [comment]
@tiran are you able to take this on?
